proposals/000-decorator-api.md

```diff
@@ -0,0 +1,322 @@
+    Decoration(
+        id: highlight.id,
+        locator: highlight.locator,
+        style: Decoration.Style.Highlight(tint: highlight.color)
+    )
+}
+
+// Finally, apply the decorations in a group named "user-highlights".
+navigator.apply(decorations: decorations, in: "user-highlights")
+```
+
+### Handling User Clicks on Decorations
+
+To allow user clicks/taps on a decoration, you must first check that the `activate` interaction is available for the decoration style:
+
+```swift
+navigator.supportsDecorationInteraction(.activate, forStyle: Decoration.Style.Highlight)
+```
+
+Then, enable the `activate` interaction when creating each `Decoration` object.
+
+```swift
+Decoration(
+    id: "12",
+    locator: Locator(...),
+    style: Decoration.Style.Highlight(),
+    interactions: [.activate]
+)
+```
+
+Finally, listen to the `DecorationActivated` event which will be emitted when the user clicks the decoration. The API depends on the platform.
+
+```swift
+navigator.observe(DecorationActivated) { decoration in
+    // Present a highlight pop-up for `decoration`...
+}
+```
+
+### Registering Decoration Styles
+
+Reading apps are welcome to register new decoration styles for custom use cases. The API depends on each Navigator, but here's an example using an HTML Navigator (e.g. for EPUB) to implement a sidemark:
+
+```swift
+// Declare a new style with associated configuration values.
+class Decoration.Style.Sidemark {
+    let tint: Color?
+}
+
+// Create a new HTML decoration template which will be used with our new style.
+let sidemarkTemplate = HTMLDecorationTemplate(
+    layout: .bounds,
+    width: .page,
+    element: { decoration in
+        let style = decoration.style as? Decoration.Style.Sidemark
+        let tint = style?.tint ?? Color.red
+        """
+        <div><div class="sidemark" style="--tint: \(tint.cssColor)"/></div>
+        """,
+    },
+    stylesheet: """
+        .sidemark {
+            float: left;
+            width: 5px;
+            height: 100%;
+            background-color: var(--tint);
+            margin-left: 20px;
+            border-radius: 3px;
+        }
+        [dir=rtl] .sidemark {
+            float: right;
+            margin-left: 0px;
+            margin-right: 20px;
+        }
+        """
+)
+
+// Associate the style with our template when creating the HTML navigator.
+var config = HTMLNavigator.Configuration()
+config.decorationStyles[Decoration.Style.Sidemark] = sidemarkTemplate
+let navigator = HTMLNavigator(config)
+```
+
+![](assets/000-sidemark.png)
+
+### Checking Whether a Navigator Supports a Decoration Style
+
+You should check whether the Navigator supports drawing the decoration styles required by a particular feature before enabling it. For example, underlining an audiobook does not make sense, so an Audiobook Navigator would not support the `underline` decoration style.
+
+```swift
+navigator.supportsDecorationStyle(Decoration.Style.Underline)
```

Comment:
> Besides, I'm not sure it is useful.

Consider a reading app handling both PDF and EPUB. Both navigators support highlight but only EPUB supports underline (hypothetically). Unless we have a way to ask the navigator which styles it supports, we don't know if we should hide or show the underline option in the selection pop-up.

Alternatively, we might show search results as underlined but fallback on highlights if not available. In this case we don't want to error out.

> This would break the ability of the navigator to adapt itself to different mediatypes.

That's true, and related to your other comment https://github.com/readium/architecture/pull/160#discussion_r665470584 I think it's fine in the current state of the Navigator, but could evolve with Navigator 2. In which case we would have this API at the level of the `ResourceAdapter`, maybe.

Reply:
> In which case we would have this API at the level of the ResourceAdapter, maybe.

Sounds good, eventually it should depend on the current resource.